probe/awsecs/client.go

```diff
@@ -31,72 +70,146 @@ func newClient(cluster string) (*ecsClient, error) {
 	}
 
 	return &ecsClient{
-		client:  ecs.New(sess, &aws.Config{Region: aws.String(region)}),
-		cluster: cluster,
+		client:       ecs.New(sess, &aws.Config{Region: aws.String(region)}),
+		cluster:      cluster,
+		taskCache:    map[string]ecsTask{},
+		serviceCache: map[string]ecsService{},
 	}, nil
 }
 
-// returns a map from deployment ids to service names
-func (c ecsClient) getDeploymentMap(services map[string]*ecs.Service) map[string]string {
-	results := map[string]string{}
-	for serviceName, service := range services {
-		for _, deployment := range service.Deployments {
-			results[*deployment.Id] = serviceName
-		}
+func newECSTask(task *ecs.Task) ecsTask {
+	now := time.Now()
+	return ecsTask{
+		taskARN:           *task.TaskArn,
+		createdAt:         *task.CreatedAt,
+		taskDefinitionARN: *task.TaskDefinitionArn,
+		startedAt:         *task.StartedAt,
+		startedBy:         *task.StartedBy,
+		fetchedAt:         now,
+		lastUsedAt:        now,
+	}
+}
+
+func newECSService(service *ecs.Service) ecsService {
+	now := time.Now()
+	deploymentIDs := make([]string, 0, len(service.Deployments))
+	for _, deployment := range service.Deployments {
+		deploymentIDs = append(deploymentIDs, *deployment.Id)
+	}
+	return ecsService{
+		serviceName:       *service.ServiceName,
+		deploymentIDs:     deploymentIDs,
+		desiredCount:      *service.DesiredCount,
+		pendingCount:      *service.PendingCount,
+		runningCount:      *service.RunningCount,
+		taskDefinitionARN: *service.TaskDefinition,
+		fetchedAt:         now,
+		lastUsedAt:        now,
 	}
+}
+
+// Returns a channel from which service ARNs can be read.
+// Cannot fail as it will attempt to deliver partial results, though that may end up being no results.
+func (c ecsClient) listServices() <-chan string {
+	log.Debugf("Listing ECS services")
+	results := make(chan string)
+	go func() {
+		count := 0
+		err := c.client.ListServicesPages(
+			&ecs.ListServicesInput{Cluster: &c.cluster},
+			func(page *ecs.ListServicesOutput, lastPage bool) bool {
+				for _, arn := range page.ServiceArns {
+					count++
+					results <- *arn
+				}
+				return true
+			},
+		)
+		if err != nil {
+			log.Warnf("Error listing ECS services, ECS service report may be incomplete: %v", err)
+		}
+		log.Debugf("Listed %d services", count)
+		close(results)
+	}()
 	return results
 }
 
-// cannot fail as it will attempt to deliver partial results, though that may end up being no results
-func (c ecsClient) getServices() map[string]*ecs.Service {
-	results := map[string]*ecs.Service{}
-	lock := sync.Mutex{} // lock mediates access to results
+// Returns (input, done) channels. Service ARNs given to input are batched and details are fetched,
+// with full ecsService objects being put into the cache. Closes done when finished.
+func (c ecsClient) describeServices() (chan<- string, <-chan bool) {
```

Comment:
It doesn't look like you are sending anything over the done channel. I would use a `chan struct{}` instead of a `chan bool`